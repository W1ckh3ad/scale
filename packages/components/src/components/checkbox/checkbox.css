.input {
  position: relative;
}
.input .input__helper-text {
  font-weight: var(--scl-font-weight-bold);
}

.input .input__helper-text {
  transition: all var(--scl-motion-duration-fast)
    var(--scl-motion-easing-default);
  padding-left: var(--scl-spacing-12);
  font-size: var(--scl-font-size-12);
  line-height: var(--scl-font-line-height-133);
  color: var(--scl-color-blue-70);
}
.input .input__meta {
  display: flex;
  justify-content: space-between;
}

.input--status-error .input__input,
.input--status-error .input__textarea,
.input--status-error .input__select {
  border: var(--scl-spacing-2) solid var(--scl-color-background-error);
}
.input--status-error .input__helper-text {
  color: var(--scl-color-text-error);
}
.input--status-error .input__counter {
  color: var(--scl-color-text-error);
}
.input--type-checkbox {
  display: flex;
  flex-wrap: wrap;
  align-items: center;
}
.input--type-checkbox input {
  width: 0;
  height: 0;
  opacity: 0;
  position: absolute;
}
.input--type-checkbox .input__meta {
  width: 100%;
}
.input--type-checkbox .input__helper-text {
  padding-left: var(--scl-spacing-32);
}
.input--type-checkbox label {
  color: var(--scl-color-text-default);
  font-weight: var(--scl-font-weight-medium);
}
.input--type-checkbox .input__checkbox-container {
  width: var(--scl-spacing-24);
  height: var(--scl-spacing-24);
  display: flex;
  position: relative;
  align-items: center;
  margin-right: var(--scl-spacing-8);
}
.input--type-checkbox input:disabled ~ label {
  /* FIXME make sure color is right */
  color: var(--scl-color-background-disabled);
}
.input--type-checkbox input:checked:disabled ~ label {
  color: var(--scl-color-background-disabled);
}
.input--type-checkbox
  input:checked:disabled
  ~ .input__checkbox-container
  .input__checkbox-placeholder {
  background: var(--scl-color-background-disabled);
}
/* TODO this could removed once scale-icon is removed */
.input--type-checkbox
  input:checked:disabled
  ~ .input__checkbox-container
  scale-icon {
  --icon-color: var(--scl-color-grey-30);
}
.input--type-checkbox
  input:checked:not([disabled]):hover
  ~ .input__checkbox-container
  .input__checkbox-placeholder,
.input--type-checkbox
  input:checked:not([disabled])
  ~ .input__checkbox-container:hover
  .input__checkbox-placeholder {
  box-shadow: none;
  border-color: var(--scl-color-primary-hover, #f90984);
  background: var(--scl-color-primary-hover, #f90984);
}
.input--type-checkbox
  input:checked:not([disabled]):active
  ~ .input__checkbox-container
  .input__checkbox-placeholder,
.input--type-checkbox
  input:checked:not([disabled])
  ~ .input__checkbox-container:active
  .input__checkbox-placeholder {
  border-color: var(--scl-color-primary-active, #cb0068);
  background: var(--scl-color-primary-active, #cb0068);
}
/* TODO this could removed once scale-icon is removed */
.input--type-checkbox
  input:checked:not([disabled]):active
  ~ .input__checkbox-container
  scale-icon,
.input--type-checkbox
  input:checked:not([disabled])
  ~ .input__checkbox-container:active
  scale-icon {
  --icon-color: var(--scl-color-white);
}
.input--type-checkbox
  input:checked:not([disabled])
  ~ .input__checkbox-container
  .input__checkbox-placeholder {
  border: var(--scl-spacing-1) solid var(--scl-color-primary, #e20074);
  background: var(--scl-color-primary, #e20074);
}
.input--type-checkbox
  input:disabled
  ~ .input__checkbox-container
  .input__checkbox-placeholder {
  border-color: var(--scl-color-background-disabled);
}
.input--type-checkbox
  input:focus
  ~ .input__checkbox-container
  .input__checkbox-placeholder {
  box-shadow: 0 0 0 var(--scl-spacing-2) var(--scl-color-focus);
}
.input--type-checkbox
  input:not([disabled]):hover
  ~ .input__checkbox-container
  .input__checkbox-placeholder,
.input--type-checkbox
  input:not([disabled])
  ~ .input__checkbox-container:hover
  .input__checkbox-placeholder {
  box-shadow: none;
  border-color: var(--scl-color-primary-hover, #f90984);
}
.input--type-checkbox
  input:not([disabled]):hover
  ~ .input__checkbox-container
  ~ label,
.input--type-checkbox
  input:not([disabled])
  ~ .input__checkbox-container:hover
  ~ label {
  color: var(--scl-color-primary-hover, #f90984);
}
.input--type-checkbox
  input:not([disabled]):active
  ~ .input__checkbox-container
  .input__checkbox-placeholder,
.input--type-checkbox
  input:not([disabled])
  ~ .input__checkbox-container:active
  .input__checkbox-placeholder {
  border-color: var(--scl-color-primary-active, #cb0068);
  background: var(--scl-color-primary-active, #cb0068);
}
/* TODO this could removed once scale-icon is removed */
.input--type-checkbox
  input:not([disabled]):active
  ~ .input__checkbox-container
  scale-icon,
.input--type-checkbox
  input:not([disabled])
  ~ .input__checkbox-container:active
  scale-icon {
  --icon-color: var(--scl-color-primary-active, #cb0068);
}
.input--type-checkbox
  input:not([disabled]):active
  ~ .input__checkbox-container
  ~ label,
.input--type-checkbox
  input:not([disabled])
  ~ .input__checkbox-container:active
  ~ label {
  color: var(--scl-color-primary-active, #cb0068);
}
.input--type-checkbox .input__checkbox-container .input__checkbox-placeholder {
  width: var(--scl-spacing-16);
  height: var(--scl-spacing-16);
  margin: var(--scl-spacing-2);
  box-sizing: border-box;
  transition: all var(--scl-motion-duration-fast)
    var(--scl-motion-easing-default);
  margin-left: var(--scl-spacing-4);
  border-radius: var(--scl-radius-4);
  border: var(--scl-spacing-1) solid var(--scl-color-text-default);
  background: var(--scl-color-background-default);
}
.input--type-checkbox .input__checkbox-container scale-icon {
  top: calc(0.5 * var(--scl-spacing-12));
  left: calc(0.5 * var(--scl-spacing-12));
  width: var(--scl-spacing-12);
  height: var(--scl-spacing-12);
  position: absolute;
  user-select: none;
  /* TODO this could removed once scale-icon is removed */
  --icon-color: var(--scl-color-background-default);
}
.input--type-checkbox .input__checkbox-container ~ label {
  transition: all var(--scl-motion-duration-fast)
    var(--scl-motion-easing-default);
}
.input--type-checkbox.input--status-error
  .input__checkbox-container
  .input__checkbox-placeholder {
  border: var(--scl-spacing-2) solid var(--scl-color-background-error);
}

.input--disabled label,
.input--disabled input,
.input--disabled .input__checkbox-container,
.input--disabled .input__helper-text {
  cursor: not-allowed;
  border-color: var(--scl-color-background-disabled);
  color: var(--scl-color-background-disabled);
  background: var(--scl-color-white);
}
