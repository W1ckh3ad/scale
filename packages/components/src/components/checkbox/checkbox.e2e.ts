import { newE2EPage } from '@stencil/core/testing';

describe('scale-checkbox', () => {
  it('renders', async () => {
    const page = await newE2EPage();
    await page.setContent('<scale-checkbox/>');
    const element = await page.find('scale-checkbox');
    expect(element).toHaveClass('hydrated');
  });
});
