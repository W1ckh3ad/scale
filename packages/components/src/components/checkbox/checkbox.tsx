import {
  Component,
  Element,
  Event,
  EventEmitter,
  h,
  Host,
  Prop,
} from '@stencil/core';
import classNames from 'classnames';

let i = 0;

@Component({
  tag: 'scale-checkbox',
  styleUrl: './checkbox.css',
  shadow: false,
})
export class Checkbox {
  @Element() el: HTMLElement;
  /** (optional) Input name */
  @Prop() name?: string = '';
  /** (optional) Input label */
  @Prop() label: string = '';
  /** (optional) Input helper text */
  @Prop() helperText?: string = '';
  /** (optional) Input status */
  @Prop() status?: string = '';
  /** (optional) Input disabled */
  @Prop() disabled?: boolean;
  /** (optional) Active switch */
  @Prop({ reflect: true }) checked?: boolean = false;
  /** (optional) Input value */
  @Prop({ mutable: true }) value?: string | number | null = '';
  /** (optional) Input checkbox id */
  @Prop() inputId?: string;
  /** (optional) Input checkbox checked icon */
  @Prop() icon?: string =
    'M9 20.215L1.695 12.91a1.25 1.25 0 011.765-1.77L9 16.68l12-12a1.25 1.25 0 011.765 1.765z';
  /** (optional) Injected CSS styles */
  @Prop() styles?: string;
  /** Emitted when the value has changed. */
  @Event() scaleChange!: EventEmitter;

  componentWillLoad() {
    if (this.inputId == null) {
      this.inputId = 'input-' + i++;
    }
  }

  render() {
    const ariaInvalidAttr =
      this.status === 'error' ? { 'aria-invalid': true } : {};
    const helperTextId = `helper-message-${i}`;
    const ariaDescribedByAttr = { 'aria-describedBy': helperTextId };

    return (
      <Host checked={this.checked}>
        <div class={this.getCssClassMap()}>
          <label class="input__checkbox-label-wrapper" htmlFor={this.inputId}>
            <input
              type="checkbox"
              name={this.name}
              id={this.inputId}
              onChange={(e: any) => {
                this.checked = e.target.checked;
                // bubble event through the shadow dom
                this.scaleChange.emit({ value: this.checked });
              }}
              value={this.value}
              checked={this.checked}
              disabled={this.disabled}
              {...ariaInvalidAttr}
              {...ariaDescribedByAttr}
            />
            <div class={classNames('input__checkbox-container')}>
              <span class={classNames('input__checkbox-placeholder')}></span>
              {/* Accessibility: rendering the icon *only* when checked, otherwise is always visible in HCM */}
              {!!this.icon && this.checked && (
                <scale-icon path={this.icon} size={12}></scale-icon>
              )}
            </div>
            <span class="input__checkbox-label">{this.label}</span>
            {!!this.helperText && (
              <div
                class="input__meta"
                id={helperTextId}
                aria-live="polite"
                aria-relevant="additions removals"
              >
                <div class="input__helper-text">{this.helperText}</div>
              </div>
            )}
          </label>
        </div>
      </Host>
    );
  }

  getCssClassMap() {
    return classNames(
      'input',
      'input--type-checkbox',
      this.checked && `input--checked`,
      this.disabled && `input--disabled`,
      this.status && `input--status-${this.status}`
    );
  }
}
