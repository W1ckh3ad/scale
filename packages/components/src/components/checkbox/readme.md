# scale-checkbox



<!-- Auto Generated Below -->


## Properties

| Property     | Attribute     | Description                            | Type               | Default                                                                                     |
| ------------ | ------------- | -------------------------------------- | ------------------ | ------------------------------------------------------------------------------------------- |
| `checked`    | `checked`     | (optional) Active switch               | `boolean`          | `false`                                                                                     |
| `disabled`   | `disabled`    | (optional) Input disabled              | `boolean`          | `undefined`                                                                                 |
| `helperText` | `helper-text` | (optional) Input helper text           | `string`           | `''`                                                                                        |
| `icon`       | `icon`        | (optional) Input checkbox checked icon | `string`           | `'M9 20.215L1.695 12.91a1.25 1.25 0 011.765-1.77L9 16.68l12-12a1.25 1.25 0 011.765 1.765z'` |
| `inputId`    | `input-id`    | (optional) Input checkbox id           | `string`           | `undefined`                                                                                 |
| `label`      | `label`       | (optional) Input label                 | `string`           | `''`                                                                                        |
| `name`       | `name`        | (optional) Input name                  | `string`           | `''`                                                                                        |
| `status`     | `status`      | (optional) Input status                | `string`           | `''`                                                                                        |
| `styles`     | `styles`      | (optional) Injected CSS styles         | `string`           | `undefined`                                                                                 |
| `value`      | `value`       | (optional) Input value                 | `number \| string` | `''`                                                                                        |


## Events

| Event         | Description                         | Type               |
| ------------- | ----------------------------------- | ------------------ |
| `scaleChange` | Emitted when the value has changed. | `CustomEvent<any>` |


## Dependencies

### Depends on

- [scale-icon](../icon)

### Graph
```mermaid
graph TD;
  scale-checkbox --> scale-icon
  style scale-checkbox fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
