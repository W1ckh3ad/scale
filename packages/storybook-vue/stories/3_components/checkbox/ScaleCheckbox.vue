<template>
  <scale-checkbox
    :name="name"
    :label="label"
    :helper-text="helperText"
    :disabled="disabled"
    :checked="checked"
    :value="value"
    :status="status"
    :input-id="inputId"
    @scaleChange="scaleChange"
  >
    <slot></slot>
  </scale-checkbox>
</template>

<script>
import { action } from "@storybook/addon-actions";

export default {
  props: {
    name: { type: String },
    label: { type: String },
    helperText: { type: String },
    disabled: { type: Boolean },
    checked: { type: Boolean },
    value: { type: String },
    status: { type: String },
    inputId: { type: String },
    styles: { type: String }
  },
  methods: {
    scaleChange($event) {
      action("scaleChange");
      this.$emit("scaleChange", $event);
    }
  }
};
</script>
