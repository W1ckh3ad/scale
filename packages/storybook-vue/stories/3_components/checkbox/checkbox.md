# Checkbox

With checkboxes, users select and deselect single options in a form.

## General

Checkboxes are used for multiple choice situations. Each checkbox works independently of the other checkboxes, therefore checking an additional checkbox does not affect other selections. Checkboxes can be selected and deselected by clicking either the box or the corresponding label.

![Image Name](/assets/3_components/checkbox/image-20200810103827627.png)

### When to Use

**Forms**<br>
Checkboxes can be used in forms on a full page, in modals or on side panels.

**Filtering and Batch Action**<br>
They can be used to filter data either on a page, in a menu or within a component.

**Terms and Conditions**<br>
Selecting or deselecting the checkbox can indicate whether you agree to the terms.

### When Not to Use

If you want the user to choose only one mandatory option from a list, use radio buttons instead.

## Elements

### Label

Give checkboxes clear labels so users know what they are selecting or deselecting e.g. "Sign up for the Newsletter".

### Required Input / Improve

When checking a box is required, use this flag to display the corresponding information and validate the form accordingly.

## Related Components

Switch, Select Box, Radio Button

## Code
