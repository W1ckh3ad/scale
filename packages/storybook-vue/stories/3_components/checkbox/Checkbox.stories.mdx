import { Meta, Story, ArgsTable, Canvas, Description } from "@storybook/addon-docs/blocks";
import ScaleCheckbox from "./ScaleCheckbox.vue";
import { action } from "@storybook/addon-actions";
export const ICON = "M21.15,4.85 L9,17.05 L3.3,11.3 C2.9,10.9 2.3,10.9 1.9,11.3 C1.5,11.7 1.5,12.3 1.9,12.7 L9,19.85 L22.6,6.25 C23,5.85 23,5.25 22.6,4.85 C22.2,4.45 21.55,4.5 21.15,4.85 Z";

<Meta
  title="Components/Checkbox"
  component={ScaleCheckbox}
  argTypes={{}}
/>

export const Template = (args, { argTypes }) => ({
  components: { ScaleCheckbox },
  props: {
    label: String,
    ...ScaleCheckbox.props
  },
  template: `
    <scale-checkbox
      :name="name"
      :label="label"
      :helper-text="helperText"
      :disabled="disabled"
      :checked="checked"
      :value="value"
      :status="status"
      :input-id="inputId"
      :icon="icon"
      @scaleChange="scaleChange"
    ></scale-checkbox>
  `,
  methods: {
    scaleChange: function() {
      action('scaleChange')();
    },
  },
})

<div style={{display: 'inline-flex', alignItems: 'center', justifyContent: 'space-between', width: '100%'}}>
    <h1>Checkbox</h1>
    <img src="assets/aa.png" alt="Accessible AA" />
</div>


## Standard

<Canvas withSource="none">
  <Story
    name="Standard"
    args={{
      label: "Checkbox",
      icon: ICON
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

<ArgsTable story="Standard" />


## Standard Disabled

<Canvas withSource="none">
  <Story
    name="Standard Disabled"
    args={{
      label: "Standard Disabled",
      icon: ICON,
      disabled: true
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-checkbox disabled @scaleChange="handleChange" label="Standard Disabled"></scale-checkbox>
```

## Selected

<Canvas withSource="none">
  <Story
    name="Selected"
    args={{
      label: "Selected",
      icon: ICON,
      checked: true
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-checkbox checked @scaleChange="handleChange" label="Selected"></scale-checkbox>
```

## Selected Disabled

<Canvas withSource="none">
  <Story
    name="Selected Disabled"
    args={{
      label: "Selected Disabled",
      icon: ICON,
      checked: true,
      disabled: true,
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-checkbox checked disabled @scaleChange="handleChange" label="Selected Disabled"></scale-checkbox>
```


## Error

<Canvas withSource="none">
  <Story
    name="Error"
    args={{
      label: "Standard with error",
      icon: ICON,
      status: "error",
      helperText: "This is the error message"
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-checkbox status="error" @scaleChange="handleChange" label="Standard with error"></scale-checkbox>
```


## Custom Label Slot

<Canvas withSource="none">
  <Story
    name="Custom Label Slot"
    args={{}}
  >
    {() => ({
        components: { ScaleCheckbox },
        props: {
            label: String,
            ...ScaleCheckbox.props
        },
        template: `<scale-checkbox><span style="cursor: pointer; display: flex; align-items: center;">Custom label &nbsp;<scale-icon-alert-imprint-dataprivacy></scale-icon-alert-imprint-dataprivacy></span></scale-checkbox>`,
    })}
  </Story>
</Canvas>

```html
<scale-checkbox><span style="cursor: pointer; display: flex; align-items: center;">Custom label &nbsp;<scale-icon-alert-imprint-dataprivacy></scale-icon-alert-imprint-dataprivacy></span></scale-checkbox>
```
