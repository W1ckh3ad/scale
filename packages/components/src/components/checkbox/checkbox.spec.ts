import { newSpecPage } from '@stencil/core/testing';
import { Checkbox } from './checkbox';
describe('Checkbox', () => {
  it('should match snapshot', async () => {
    const page = await newSpecPage({
      components: [Checkbox],
      html: `<scale-checkbox></scale-checkbox>`,
    });
    expect(page.root).toMatchSnapshot();
  });
});
