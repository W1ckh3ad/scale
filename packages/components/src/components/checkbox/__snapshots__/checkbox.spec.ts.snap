// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox should match snapshot 1`] = `
<scale-checkbox>
  <div class="input input--type-checkbox">
    <input aria-describedby="helper-message-1" id="input-0" type="checkbox" value="">
    <div class="input__checkbox-container">
      <span class="input__checkbox-placeholder"></span>
    </div>
    <label htmlfor="input-0"></label>
  </div>
</scale-checkbox>
`;
